### ISSUE TITLE
Add Support for Top-level await

### ISSUE THREAD

**Issue by NemoStein** [eslint/eslint#14632]
Today Top-level await reached Stage 4.
This brings new depth to async modules, and I'm looking forward to seeing it implemented in ESLint in the near future.

- [x] Ecma262 update: https://github.com/tc39/ecma262/pull/2408
- [x] ESTree update: n/a
- [x] Acorn update: https://github.com/acornjs/acorn/pull/1040
- [x] `eslint-visitor-keys` 
- [x] `espree` update: https://github.com/eslint/espree/pull/505
- [x] `eslint-scope` update: 
- [ ] `eslint` update:

**Comment by janosh** [comment]
For reference, the relevant comment in the [tc39 proposal](https://github.com/tc39/proposal-top-level-await/issues/113#issuecomment-848189231).

**Comment by ljharb** [comment]
(The canonical source of truth is here: https://github.com/tc39/proposals/commit/99fa42a0863b1915eccead5bd2939ed5f468caab)

**Comment by nzakas** [comment]
Thanks for opening this. We can definitely start implanting it and would love some help to get started.

**Comment by ota-meshi** [comment]
I opened a pull request in acorn.
https://github.com/acornjs/acorn/pull/1040

**Comment by Slapbox** [comment]
Related (but old and outdated) https://github.com/eslint/eslint/issues/13178

**Comment by nimit2801** [comment]
What is the status of this feature, I want to use eslint with my project and want a way to wave of the error?
Can someone suggest to me what should I do from the following
```
1. Add rules or something to ignore this error if yes any examples
2. If this is supported what things should I mention in `.eslintrc.json` to use this feature
```
Question: what version of ecma does top-level await joins in?
Thank You

**Comment by saschanaz** [comment]
Probably blocked by #14591?

**Comment by nzakas** [comment]
Yup. Support will be released in v8.0.0. We are finishing up that work now.

**Comment by ktx-mega** [comment]
Do you have an idea of the release date ?

**Comment by nzakas** [comment]
Nothing firm. We will likely release a beta within the next two weeks and the final version will depend on the feedback we get during the beta.

**Comment by joshunger** [comment]
Happy afternoon (or morning)! Will this fix `Parsing error: Can not use keyword 'await' outside an async function` and what specific 8.0 beta version? I just tried `v8.0.0-beta.1` with and mjs file and it did not work. Thanks.

**Comment by nzakas** [comment]
@joshunger it should. Be sure you have ecmaVersion set to 2022 and sourceType set to module. If you’re still having trouble, please stop by the [chat](https://eslint.org/chat/help).

**Comment by joshunger** [comment]
@nzakas thanks for the help I missed the `ecmaVersion`. It is working great!

**Comment by AndreiSoroka** [comment]
Hi!

```bash
> eslint . --ext .mjs --fix


Oops! Something went wrong! :(

ESLint: 8.0.0-beta.1

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/util/glob-util' is not defined by "exports" in /Users/andreisoroka/git/other/faststartservice/node_modules/eslint/package.json
    at throwExportsNotFound (internal/modules/esm/resolve.js:285:9)
    at packageExportsResolve (internal/modules/esm/resolve.js:508:3)
    at resolveExports (internal/modules/cjs/loader.js:450:36)
    at Function.Module._findPath (internal/modules/cjs/loader.js:490:31)
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:888:27)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (/Users/andreisoroka/git/other/faststartservice/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/Users/andreisoroka/git/other/faststartservice/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:43:34)
    at Module._compile (/Users/andreisoroka/git/other/faststartservice/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! fuststartservice@0.0.0 lint:fix: `eslint . --ext .mjs --fix`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the fuststartservice@0.0.0 lint:fix script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

package.json
```json5
{
  "scripts": {
    "lint": "eslint . --ext .mjs",
    "lint:fix": "eslint . --ext .mjs --fix"
  },
  "devDependencies": {
    // "eslint": "^7.32.0", // was
    "eslint": "^8.0.0-beta.1",
    "eslint-config-airbnb-base": "^14.2.1",
    "eslint-plugin-import": "^2.24.2"
  }
}
```

```json5
{
  "env": {
    // "es2021": true, (and with 2021 don't work)
    "es2022": true,
    "node": true
  },
  "extends": [
    "airbnb-base"
  ],
  "parserOptions": {
    // "ecmaVersion": 12, // was
    "ecmaVersion": 2022,
    "sourceType": "module"
  },
  "settings": {
    "import/resolver": {
      "node": {
        "extensions": [
          ".js",
          ".mjs"
        ]
      }
    }
  },
  "rules": {
    "no-console": "off",
    "no-underscore-dangle": "off",
    "import/extensions": [
      "error",
      "always"
    ]
  }
}

```

**Comment by joshunger** [comment]
@AndreiSoroka Hello! Try removing `eslint-config-airbnb-base`.

**Comment by mdjermanovic** [comment]
Hi @AndreiSoroka!

> ```shell
> ESLint: 8.0.0-beta.1
> 
> Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/util/glob-util' is not defined by "exports" in /Users/andreisoroka/git/other/faststartservice/node_modules/eslint/package.json
>     at throwExportsNotFound (internal/modules/esm/resolve.js:285:9)
>     at packageExportsResolve (internal/modules/esm/resolve.js:508:3)
>     at resolveExports (internal/modules/cjs/loader.js:450:36)
>     at Function.Module._findPath (internal/modules/cjs/loader.js:490:31)
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:888:27)
>     at Function.Module._load (internal/modules/cjs/loader.js:746:27)
>     at Module.require (internal/modules/cjs/loader.js:974:19)
>     at require (/Users/andreisoroka/git/other/faststartservice/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
>     at Object.<anonymous> (/Users/andreisoroka/git/other/faststartservice/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:43:34)
> ```

This is a problem in `eslint-plugin-import`, it's using some internal APIs that have been removed in ESLint 8.

Can you please report this problem to https://github.com/import-js/eslint-plugin-import ?

Thanks for testing ESLint prereleases and let us know if you find any other problems! Just please open separate issue(s) so we could track them, this one is only about adding support for top-level await.

**Comment by AndreiSoroka** [comment]
Thanks! I check issues and find https://github.com/import-js/eslint-plugin-import/issues/2211

**Comment by AndreiSoroka** [comment]
FYI: @ljharb

**Comment by ljharb** [comment]
I’m aware; most of the ecosystem doesn’t support eslint 8 yet, which is fine since it’s not actually out yet.

**Comment by nzakas** [comment]
Please open new issues if you find problems with ESLint 8 functionality.
